src/main/java/it/pagopa/transactions/commands/handlers/TransactionRequestAuthorizationHandler.java

```diff
@@ -47,47 +49,41 @@ public Mono<RequestAuthorizationResponseDto> handle(TransactionRequestAuthorizat
             return Mono.error(new AlreadyProcessedException(transaction.getRptId()));
         }
 
-        return paymentGatewayClient.requestGeneralAuthorization(command.getData())
-                .flatMap(authResponse -> {
-                    log.info("Logging authorization event for rpt id {}", transaction.getRptId().value());
+        var monoPostePay = Mono.just(command.getData())
+                .flatMap(d -> paymentGatewayClient.requestPostepayAuthorization(d))
+                .switchIfEmpty(Mono.empty())
+                .map(p -> Tuples.of(p.getRequestId(),p.getUrlRedirect()));
 
-                    if(!authResponse.getT1().isPresent() && !authResponse.getT2().isPresent()) {
-                        return Mono.error(new BadRequestException("No gateway matched"));
-                    }
+        var monoXPay = Mono.just(command.getData())
+                .flatMap(d -> paymentGatewayClient.requestXPayAuthorization(d))
+                .switchIfEmpty(Mono.empty())
+                .map(p -> Tuples.of(p.getRequestId(),p.getUrlRedirect()));
 
-                    Mono<Tuple2<String,String>> monoPostePay = Mono.just(authResponse.getT1()).filter(pPayAuth -> pPayAuth.isPresent())
-                            .switchIfEmpty(Mono.empty())
-                            .flatMap(pPay -> Mono.zip(Mono.just(pPay.get().getRequestId()),Mono.just(pPay.get().getUrlRedirect())));
+        return monoPostePay.switchIfEmpty(monoXPay).switchIfEmpty(Mono.error(new BadRequestException("No gateway matched")))
```

Comment:
What about creating a list of `Mono`s and reducing over that?
Something like:

```java
List<Mono<Tuple<String, String>>> gatewayRequests = List.of(monoPostePay, monoXPay);

Mono<Tuple2<String, String>> gatewayAttempts = gatewayRequests.stream().reduce((pipeline, candidateStep) -> {
    return pipeline.switchIfEmpty(candidateStep);
});

gatewayAttempts.switchIfEmpty(Mono.error(new BadRequestException("No gateway matched")));
```

Reply:
That's a code emotion 💘